### PR TITLE
feat(condo): NO-JIRA ip black list support

### PR DESCRIPTION
**Pull request by ekabardinsky**

### CODE REVIEW

File: packages/keystone/ipBlackList.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+const express = require('express')
+const { isEmpty } = require('lodash')
+
+const conf = require('@open-condo/config')
+const { getIp } = require('@open-condo/keystone/ip.utils')
+
+const ipBlackList = (conf.IP_BLACK_LIST || '').split(',')
```
[the remark this answers is not in the document]
Comment (ekabardinsky) [reply packages/keystone/ipBlackList.js:7]: I'll put in .werf a little bit later

**Review by sitozzz — APPROVED** [review]

File: apps/condo/domains/user/schema/ConfirmPhoneActionService.js [hunk 2]
```diff
@@ -106,8 +107,8 @@ const ipWhiteList = conf.IP_WHITE_LIST ? JSON.parse(conf.IP_WHITE_LIST) : []
 const maxSmsForIpByDay = Number(conf['MAX_SMS_FOR_IP_BY_DAY']) || MAX_SMS_FOR_IP_BY_DAY
 const maxSmsForPhoneByDay = Number(conf['MAX_SMS_FOR_PHONE_BY_DAY']) || MAX_SMS_FOR_PHONE_BY_DAY
 
-const checkSMSDayLimitCounters = async (phone, rawIp) => {
-    const ip = rawIp.split(':').pop()
+const checkSMSDayLimitCounters = async (phone, req) => {
+    const ip = getIp(req)
```
Comment (pahaz) [inline apps/condo/domains/user/schema/ConfirmPhoneActionService.js:111]: `req.ip` has the right value.

File: packages/keystone/ipBlackList.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+
+class KeystoneIpBlackListMiddleware {
+    async prepareMiddleware ({ keystone }) {
+        // nosemgrep: javascript.express.security.audit.express-check-csurf-middleware-usage.express-check-csurf-middleware-usage
+        const app = express()
+        app.use((req, res, next) => {
+            const ip = getIp(req)
+            if (ipBlackList[ip]) {
+                res.status(429).send('Too Many Requests')
```
Comment (pahaz) [inline packages/keystone/ipBlackList.js:23]: It's better to use [410](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/410)

File: packages/keystone/ipBlackList.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+const ipBlackList = (conf.IP_BLACK_LIST || '').split(',')
+    .map(ip => ip.trim())
+    .filter(ip => !isEmpty(ip))
+    .reduce((map, ip) => {
+        map[ip] = true
+        return map
+    }, {})
+
+
+class KeystoneIpBlackListMiddleware {
```
Comment (pahaz) [inline packages/keystone/ipBlackList.js:16]:
```suggestion
class IpBlackListMiddleware {
```

File: apps/condo/domains/user/schema/ConfirmPhoneActionService.js (hunk 2 above)
Comment (ekabardinsky) [reply apps/condo/domains/user/schema/ConfirmPhoneActionService.js:111]: ok

File: packages/keystone/ipBlackList.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+
+class IpBlackListMiddleware {
+    async prepareMiddleware ({ keystone }) {
+        // nosemgrep: javascript.express.security.audit.express-check-csurf-middleware-usage.express-check-csurf-middleware-usage
+        const app = express()
+        app.use((req, res, next) => {
+            const ip = getIp(req)
+            if (ipBlackList[ip]) {
+                res.status(410).send('Too Many Requests')
```
Comment (AleX83Xpert) [inline packages/keystone/ipBlackList.js:23]:
```suggestion
                res.status(410).send('Gone')
```

**Review by AleX83Xpert — APPROVED** [review]

File: apps/condo/domains/user/schema/ConfirmPhoneActionService.js [hunk 6]
```diff
@@ -106,8 +107,8 @@ const ipWhiteList = conf.IP_WHITE_LIST ? JSON.parse(conf.IP_WHITE_LIST) : []
 const maxSmsForIpByDay = Number(conf['MAX_SMS_FOR_IP_BY_DAY']) || MAX_SMS_FOR_IP_BY_DAY
 const maxSmsForPhoneByDay = Number(conf['MAX_SMS_FOR_PHONE_BY_DAY']) || MAX_SMS_FOR_PHONE_BY_DAY
 
-const checkSMSDayLimitCounters = async (phone, rawIp) => {
-    const ip = rawIp.split(':').pop()
+const checkSMSDayLimitCounters = async (phone, req) => {
```
Comment (pahaz) [inline apps/condo/domains/user/schema/ConfirmPhoneActionService.js:110]: It's better to rollback this changes. It's hard to test it

File: packages/keystone/ip.utils.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
+const { isEmpty, get } = require('lodash')
+
+function getIp (req) {
```
Comment (pahaz) [inline packages/keystone/ip.utils.js:3]: it's better not to create any function. just use req.ip

File: packages/keystone/ipBlackList.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    async prepareMiddleware ({ keystone }) {
+        // nosemgrep: javascript.express.security.audit.express-check-csurf-middleware-usage.express-check-csurf-middleware-usage
+        const app = express()
+        app.use((req, res, next) => {
+            const ip = getIp(req)
+            if (checkIpInDictionary(ipDictionary, ip)) {
+                res.status(410).send('Gone')
+            } else {
+                next()
+            }
```
Comment (pahaz) [inline packages/keystone/ipBlackList.js:50]:
```suggestion
            if (checkIpInDictionary(ipDictionary, req.ip)) {
                res.status(410).send('Gone')
            } else {
                next()
            }
```

**Pull request merged**